### ISSUE TITLE
require-atomic-updates false positive

### ISSUE THREAD

**Issue by medikoo** [eslint/eslint#11899]
**Tell us about your environment**

* **ESLint Version:** 6.0.1
* **Node Version:** 12.4.0
* **npm Version:** 6.9.0

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

https://github.com/medikoo/eslint-config-medikoo/blob/master/index.js

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```javascript
const fn = async someObj => {
  let { foo } = someObj;
  await Promise.resolve();
  // Possible race condition: `someObj.bar` might be reassigned based on an outdated value of `someObj.bar`
  someObj.bar = "whatever";
};
```

**What did you expect to happen?**

No error reported

**What actually happened? Please include the actual, raw output from ESLint.**

Error reported:
```
Possible race condition: `someObj.bar` might be reassigned based on an outdated value of 
```

**Are you willing to submit a pull request to fix this bug?**

Not at this point

**Comment by mysticatea** [comment]
Thank you for your report.

I think this is intentional behavior because that `someObj` object was read before `await` expression and written after `await` expression.

**Comment by medikoo** [comment]
> I think this is intentional behavior because that someObj object was read before await expression and written after await expression.

Ok, but that's pretty aggressive assumption, as in many cases that can be an intentional well-thought construct

In place where it reported for me, it's totally intentional and there's no way to write it differently (aside of dropping async/await), due to that I needed to turn off this rule globally.

Still I'd love it to report for me cases as `count += await getCount()` which are error-prone obviously (and even if not, easy to write in a way they do not report)

**Comment by stinovlas** [comment]
I would say that if nothing else, error message is pretty misleading. 
```
someObj.bar might be reassigned based on an outdated value of someObj.bar
```
That is simply not true, because `someObj.bar = "whatever";` assigns string `"whatever"` which is not depending on `someObj.bar` in any way.

Also, it's not following the [rule definition](https://eslint.org/docs/rules/require-atomic-updates#rule-details) which clearly states that following must be true:
```
A variable or property is reassigned to a new value which is based on its old value.
```

**Comment by jwalton** [comment]
Here's a pretty vanilla Express middleware that triggers this problem, as you might see in any app that does virtual hosting based on the URL, if you're looking for a concrete example:

```js
export function populateAccount() {
    return function(req, res, next) {
        Promise.resolve()
            .then(async () => {
                const host = req.headers['host'];
                req.account = await getAccount(host);
            })
            .then(() => next(), next);
    };
}
/*eslint require-atomic-updates: error */
```

[Demo link](https://eslint.org/demo/#eyJ0ZXh0IjoiZXhwb3J0IGZ1bmN0aW9uIHBvcHVsYXRlQWNjb3VudCgpIHtcclxuICAgIHJldHVybiBmdW5jdGlvbihyZXEsIHJlcywgbmV4dCkge1xyXG4gICAgICAgIFByb21pc2UucmVzb2x2ZSgpXHJcbiAgICAgICAgICAgIC50aGVuKGFzeW5jICgpID0+IHtcclxuICAgICAgICAgICAgICAgIGNvbnN0IGhvc3QgPSByZXEuaGVhZGVyc1snaG9zdCddO1xyXG4gICAgICAgICAgICAgICAgcmVxLmFjY291bnQgPSBhd2FpdCBnZXRBY2NvdW50KGhvc3QpO1xyXG4gICAgICAgICAgICB9KVxyXG4gICAgICAgICAgICAudGhlbigoKSA9PiBuZXh0KCksIG5leHQpO1xyXG4gICAgfTtcclxufVxyXG4vKmVzbGludCByZXF1aXJlLWF0b21pYy11cGRhdGVzOiBlcnJvciAqLyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OCwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==).

**Comment by j0h** [comment]
@mysticatea could the following POC also be related to the issue?

```javascript
function dataGetter() {
  return {};
}

function anotherDataGetter() {
  return {};
}

async function main() {
  let get_data = await dataGetter();

  const get_another_data = await anotherDataGetter(get_data.id);

  (() => {
    get_data[get_another_data.foo] = 1;
  });

  // Error Here: Possible race condition: `get_data` might be reassigned based on an outdated value of `get_data`.
  get_data = {
    ...get_data,
    ...get_another_data,
  };
}

main();
```

[Demo Link](https://eslint.org/demo/#eyJ0ZXh0IjoiZnVuY3Rpb24gZGF0YUdldHRlcigpIHtcbiAgcmV0dXJuIHt9O1xufVxuXG5mdW5jdGlvbiBhbm90aGVyRGF0YUdldHRlcigpIHtcbiAgcmV0dXJuIHt9O1xufVxuXG5hc3luYyBmdW5jdGlvbiBtYWluKCkge1xuICBsZXQgZ2V0X2RhdGEgPSBhd2FpdCBkYXRhR2V0dGVyKCk7XG5cbiAgY29uc3QgZ2V0X2Fub3RoZXJfZGF0YSA9IGF3YWl0IGFub3RoZXJEYXRhR2V0dGVyKGdldF9kYXRhLmlkKTtcblxuICAoKCkgPT4ge1xuICAgIGdldF9kYXRhW2dldF9hbm90aGVyX2RhdGEuZm9vXSA9IDE7XG4gIH0pKCk7XG5cbiAgZ2V0X2RhdGEgPSB7XG4gICAgLi4uZ2V0X2RhdGEsXG4gICAgLi4uZ2V0X2Fub3RoZXJfZGF0YSxcbiAgfTtcbn1cblxubWFpbigpOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

**Comment by ololoepepe** [comment]
There is a workaround:
```js
// Given:
const obj = {foo: 'bar'};

// Instead of:
obj.prop = 'whatever';

// Do:
Object.assign(obj, {prop: 'whatever'});
```
This way the rule is not triggered (as expected).
But here comes another question: given current (incorrect) behavior, shouldn't the rule be triggered on `Object.assign` as well?

**Comment by Lonniebiz** [comment]
It is my understanding that await prevents any further execution from happening, **within that code block** until the promise is resolved. `someObj` is an argument passed into that function by reference and `someObj`  is accessible everywhere it is used within that code block originally posted.

Even if you awaited a promise that did 100 things to the same object that `someObj`  references, once that promise-work is done there is nothing to stop this code block from also setting `someObj.bar = "whatever";`. Sure, the fulfilled promise could have made `AnotherReferenceTosomeObj.bar = "something specific";`, but that **should** get over written by `someObj.bar = "whatever";`

To assume that the promise does this on a **"might"**, and invoke an eslint error, seems like an attempt to oversimplify async programming to me.

If eslint isn't sophisticated enough to point out BOTH lines of code, that are "racing" to overwrite something (race condition), then it probably needs its might-suspicions to itself (until it gains that sophistication). Don't block my build on a "might". It takes two to race. If I've got a race condition, me both lines of code that are racing.

Here's my example:
https://github.com/eslint/eslint/issues/11967

**Comment by ololoepepe** [comment]
@Lonniebiz what if the other line of code is located somewhere in a 3rd-party lib? Looking through entire node_modules to find that line does no sound as a good idea.
Still, I agree with you. That suspicion should not be considered an error. A warning would be quite enough.

**Comment by ronkorving** [comment]
This is a real issue (I'm hitting all the examples pasted above), and one that was **not** mentioned in the migration guide for ESLint 6.0.0.

I can appreciate theories on how these cases *are* at least on paper unsafe, but that kind of behavior should then at least exist behind a flag. Right now, we're losing a valuable feature because the only way to deal with this situation is to turn the rule *off*.

**Comment by j0h** [comment]
> @Lonniebiz what if the other line of code is located somewhere in a 3rd-party lib? Looking through entire node_modules to find that line does no sound as a good idea.

I think that might be a non-issue, I don't think you should be setting up your eslint to run through entire node_modules. This would be outside your context/source code and it would be up to the original maintainer  of that said library to fix those issues (or yourself without an appropriate PR).

**Comment by viktor-ku** [comment]
We have to disable this rule in our company due to many false positives. In paper rule is great and should prevent many bugs, but feels like it's not quite ready yet.

**Comment by Lonniebiz** [comment]
> Still, I agree with you. That suspicion should not be considered an error. A warning would be quite enough.

I'm not even sure you should **warn** about something so [obvious and likely](https://github.com/eslint/eslint/issues/11967#issuecomment-509444223).

**Comment by ronkorving** [comment]
@mysticatea May I ask what the plan is? Will the behavior of this rule be reverted, or otherwise? Current behavior is keeping people from upgrading to ESLint 6. It would be good to know what the official stance is. If it's "won't fix", then we'll just all turn off this rule and move on with our lives. If it's "will fix in 6.0.2", then we know we can wait for a bit, and it will resolve itself.

**Comment by Lonniebiz** [comment]
@ronkorving : I'm not sure which thread is official, but did your read this?
https://github.com/standard/standard/issues/1314#issuecomment-510193785

**Comment by ralfstx** [comment]
Here is a perfectly valid koa snippet that is faulted by this rule:
```js
router.post('/webhook', async (ctx) => {
  await processRequest(ctx.request.body);
  ctx.body = 'OK';
});
```

**Comment by bcoe** [comment]
this rule is triggering in a variety of scenarios on our repos that come up frequently in samples and tests, e.g.,

```js
const _log = console.log;
const value = await Promise.resolve(10);
console.log = _log;
```

I'm in agreement with @ronkorving that this would be much more valuable as a more specific rule along the lines of, don't do this:

```js
count += await somePromise
```

**Comment by togakangaroo** [comment]
Just to add to the pile, here is one more very-common way of writing middleware that would trigger this

```
    const createNote = async (context) => {
        const { req: { entityId, entityType, note } } = context;
        const id = await notesModel.create(note, entityId, entityType);
        context.res = { id }; 
    }

```

**Comment by dzek69** [comment]
If the error message reports that value was reassigned basing on previous value, but I am just overwriting a property without using any external values - then something is definitely wrong.

This rule wasn't triggering false positives for me before 6.x.

**Comment by lvkins** [comment]
This fails for React refs too:

```
const ref = useRef(false);

async function action() {
    ref.current = false;

    await something;

    ref.current = true; // Possible race condition
}
```

**Comment by Morgul** [comment]
Just thought I'd say I was bit by this too.

Every project where I use `express` or `knex` and async/await triggered this. In none of the cases am I actually use an outdated value, I'm overriding a default value based on the return of an `await`.

**Comment by sbrth** [comment]
This is also triggered when you simply want to reset a variable.

```js
let myClass = new MyClass('hi');

const result = await myClass.doAThing();
console.log(result);

myClass = new MyClass('hi');
```

The example doesn't make sense because the real-world thing involves event triggers and a quasi-loop. I simply want to overwrite the "used-up" object with a new one for the next invocation, but for some reason this is a race condition. The new assignment doesn't involve anything about the old assignment...

**Comment by frederikhors** [comment]
I have a doubt about this code and this rule.

This is not good for eslint:

```js
let controllerLogic

export default class {
  async start () {

    if (!controllerLogic) {
      const logic = await import('./controller.js')
      controllerLogic = logic
    }

    controllerLogic.go()

  }
}
```

This one is good for eslint:

```js
let controllerLogic

export default class {
  async start () {

    if (!controllerLogic) {
      await import('./controller.js').then(logic => {
        controllerLogic = logic
      })
    }

    controllerLogic.go()

  }
}
```

## WHY?

**Comment by drahnieR** [comment]
Also, i'll give one of the cases i found here.
doSomething() just read from elem. but if the async call doesn't involve elem at all, lint error is gone.

```
async function demo(someArray) {
  return Promise.all(someArray.map(async elem => {
    const result = await doSomething(elem);	
    if (result) {
      elem.f1 = result.f1;
    }
    return elem
  }))
}
```
[Demo](https://eslint.org/demo#eyJ0ZXh0IjoiYXN5bmMgZnVuY3Rpb24gZGVtbyhzb21lQXJyYXkpIHtcbiAgcmV0dXJuIFByb21pc2UuYWxsKHNvbWVBcnJheS5tYXAoYXN5bmMgZWxlbSA9PiB7XG4gICAgY29uc3QgcmVzdWx0ID0gYXdhaXQgZG9Tb21ldGhpbmcoZWxlbSk7XHRcbiAgICBpZiAocmVzdWx0KSB7XG4gICAgICBlbGVtLmYxID0gcmVzdWx0LmYxO1xuICAgIH1cbiAgICByZXR1cm4gZWxlbVxuICB9KSlcbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7InJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOjJ9LCJlbnYiOnsiZXM2Ijp0cnVlLCJub2RlIjp0cnVlfX19)

**Comment by stinovlas** [comment]
It seems that there are enough examples of this rule not working as intended (or at least not working as described in docs). I'd say that the discussion went long enough for the maintainers to chip in and give us some comment or resolution.

I think there are several possible courses of action:
1. Fix the rule to behave as described in the docs.
1. Fix the docs to describe the actual rule behaviour.
1. Disable this rule by default.
1. Remove this rule altogether.

@mysticatea Would you be so kind and comment on this?

**Comment by jyrkive** [comment]
I believe that 3) at least would make sense. The rule in its current form flags way too much completely fine code.

**Comment by medikoo** [comment]
I think first straightforward and important step, is to remove it from `eslint:recommended` group.

Such _controversial_ rule, which doesn't work with many typical cases, should not be _recommended_ (IMO we can consider it a bug, that it landed there)

**Comment by mysticatea** [comment]
I agree that `eslint:recommended` should not contain that rule because of too aggressive.

**Comment by kaicataldo** [comment]
Agreed with @mysticstea. I’m not very familiar with this rule, but it sounds like we should consider doing the following:

1. Remove `require-atomic-updates` from `eslint:recommended`. We should be fine to release this in a minor release because it will result in fewer errors being reported.
1. Update the rule so that it’s not so aggressive with what it flags and improve documentation around it’s behavior. If this is something we can’t enforce without false positives, I think we should reconsider the rule.

**Comment by kaicataldo** [comment]
### TSC Summary

`require-atomic-updates` is currently on in `eslint:recommended` and appears to be flagging safe code (see examples in comments above). 

### TSC Question

Can we remove this rule from `eslint:recommended` while we decide how we want to move forward with the rule? This should be fine to do in a semver-minor release because it will result in fewer errors being reported.

**Comment by ronkorving** [comment]
The old implementation did add value, so I wish we could have that again.
You could move the aggressive behavior behind an option.

**Comment by Jessidhia** [comment]
> I have a doubt about this code and this rule.
> 
> This is not good for eslint:
> 
> ```js
> let controllerLogic
> 
> export default class {
>   async start () {
> 
>     if (!controllerLogic) {
>       const logic = await import('./controller.js')
>       controllerLogic = logic
>     }
> 
>     controllerLogic.go()
> 
>   }
> }
> ```
> 
> This one is good for eslint:
> 
> ```js
> let controllerLogic
> 
> export default class {
>   async start () {
> 
>     if (!controllerLogic) {
>       await import('./controller.js').then(logic => {
>         controllerLogic = logic
>       })
>     }
> 
>     controllerLogic.go()
> 
>   }
> }
> ```
> 
> ## WHY?

They are _both_ bad, but it's a bug in the rule that it cannot detect the second one as a problem.

The assignment is safe if and only if it is the only assignment done to the value AND the expression evaluated is idempotent. `await import()` happens to be, but the lint rule doesn't know that.

Why? Because the update is not atomic. It is possible that, between the `await` starting and it settling, the value of `controllerLogic` changed, which makes the premise of the branch (that `controllerLogic` is falsy) now invalid. You need to do another compare-and-swap operation instead of assuming the value it had before `await` is the same as it has after. 

For example, while `controllerLogic` is `undefined`, `.start()` is called multiple times before emptying the stack, there will be multiple assignments to `controllerLogic` pending resolution of the `await` for the result of the `import()` call. _We humans_ know that `import()` will always return the same value for the same module (presuming no hot-module-reloading related bugs); but if each expression yielded a different value, you would have three different values for `controllerLogic` and `controllerLogic.go`, and the order they would execute is unpredictable (well, it is FIFO in the case of a normal `import()` implementation, but this is information we have that the lint rule does not).

---

This response is perhaps overly pedantic, especially for a non-parallel language like (most uses of) javascript, but this is the kind of thing you need to keep in mind when writing concurrent code; this is why `Atomics` exist, this is why atomic machine language instructions exists, this is why `stdatomic.h` exists.

**Comment by ololoepepe** [comment]
@Jessidhia it is still us humans who write code in 2019. We know better what we are doing (well, not all of us, but those who don’t hardly use linters), so the rule is too aggressive and should not be enabled by default in the form it is implemented now. It is kind of if some rule gave an error if we use (or do not use) some pattern, say, factory method. It is too opinion-based for an (not perfect) algorithm.

**Comment by mikecbrant** [comment]
+1 (replying so I can watch issue)

**Comment by jwalton** [comment]
@mikecbrant If you just want to subscribe to an issue, at the very top of this page, on the right hand side, under "assignees", "labels", etc... there's a "subscribe" button that will notify you of updates to the issue:

![Subscribe Button](https://user-images.githubusercontent.com/1771003/66231066-8a93fe80-e6b3-11e9-8ede-6528818c12a7.png)

Instead of +1ing the issue, please give it a "thumbs up" at the top, so the rest of us don't all get a notification that you want to watch this issue.  :)

**Comment by platinumazure** [comment]
Per the [October 10 TSC Meeting](https://github.com/eslint/tsc-meetings/blob/master/notes/2019/2019-10-10.md#require-atomic-updates-false-positive), we intend to take the following actions:

1. Update our semver policy to allow relaxing `eslint:recommended` in a semver-minor release (currently all changes to `eslint:recommended` are semver-major).
1. Remove `require-atomic-updates` from `eslint:recommended` in the next minor release.

Of course, we still need to find a way to *fix* the rule, but hopefully this will reduce pain short-term.

**Comment by Morgul** [comment]
@platinumazure Thank you for the update, and thank you to the TSC for making these changes!

**Comment by ronkorving** [comment]
@platinumazure Thanks for the progress, that's great. May I ask why reverting the change to the `require-atomic-updates` rule was not under consideration? After that, the next step would be to fix it, and only then push that change. It's nice that `eslint:recommended` is "fixed", but it's still a broken rule.

**Comment by platinumazure** [comment]
> May I ask why reverting the change to the `require-atomic-updates` rule was not under consideration?

It just wasn't on the meeting agenda, that's all. No ulterior motive here, as far as I know. Maybe @mysticatea can speak to what benefits the most recent change to the rule brought and can offer thoughts on whether reverting that change could make sense.

**Comment by ericmorand** [comment]
I just got a build stuck because of require-atomic-updates even though the code works perfectly and I disabled this rule. It doesn't make sense to me. This is not eslint responsibility to try to guess that something may not work as expected. This is the reponsibility of the test suite.

Here we have a rule that throw an error saying:

`**Possible** race condition...`

Possible. Eslint is not sure for a good reason: it can't be sure because only a test suite can be sure that a code works as expected.

More generally, there seems to be a lot of rules that try to guess that a piece of code will work as expected. Why is eslint doing this at all? Why trying to guess things that are anyway checked by the test suite?

**Comment by ronkorving** [comment]
@ericmorand IMHO that *is* the responsibility of ESLint. It is to discourage use of patterns that are known to cause bugs. Example: you can omit `break` in a switch statement's case, and this is a language feature, but is usually not what you want it to do. Omitting it may imply a bug. While testing may expose this bug, that is irrelevant to ESLint.

Why? This philosophy of reporting possible bugs (based on patterns that are considered dangerous) goes way back in the history of linting (pre-JavaScript). ESLint did not start this. So perhaps your expectations of ESLint are not in line with its philosophy? That is fine of course, but then perhaps consider using a different tool.

Just my 2 cents.

**Comment by dzek69** [comment]
My 2 cents: all "possible" rules should exist, because usually they aren't too much about guessing but about preventing bad practices and popular errors.

If you can fix the reported error by making your code better (no "hacks", just really better) - then the rule is good.
If you need to silence the rule, because the code is all good and the only alternative is to write hacky code to fool the linter - then the rule is bad.

So currently I have nothing against other rules, it's just `require-atomic-updates` that went crazy after the update.

And the bonus thing: You can always change your config and disable all the "possible" rules if they are annoying to you 👍

**Comment by gwardwell** [comment]
Since this issue is now closed, I guess what we’re left with is turning this rule off in order to stop the false positives. If you, the ESLint team, won’t fix the bugs, then you should consider turning it off by default and making it opt-in or removing the rule altogether. 🤷‍♂️

**Comment by jwalton** [comment]
You can always submit a PR to fix it, or a PR to remove it from recommended.

**Comment by gwardwell** [comment]
@jwalton fair enough. I think I’ll take the route of removing from recommended when I can make time for a PR. I’m going to assume it works for some people, so it should remain.

**Comment by kaicataldo** [comment]
Sorry, folks - this should not have been closed. The TSC made a determination to [remove this rule from `eslint:recommended`](https://github.com/eslint/eslint/issues/11899#issuecomment-541714523), and it looks like we forgot to update labels. I'll make a PR for this right now.

**Comment by tranvansang** [comment]
I do not think this should be considered as a bug, it is a true positive warning, IMHO.

With the code snippet posted in OP, two different calls of `fn` with a same `someObj` do not mean the last value stayed in `someObj.bar` is assigned by the latter call of `fn`. Which explains why this pattern produces a race condition.

I have written [a blog post](https://transang.me/promise-based-semaphore/) about promise base semaphore pattern. This pattern can help avoiding this race condition effectively.

**Comment by JoshCoady** [comment]
> Right now, we're losing a valuable feature because the only way to deal with this situation is to turn the rule _off_.

That's not the only option. You could just switch it to a warning:

```js
  rules: {
    'require-atomic-updates': 1,
  },
```

It _does_ warn about a valid issue that wont always crop up, but when it does, it is the type of insidious bug that could take days to discover. There are quite a few comments in this thread that belie the commenter's understanding of parallel programming.

I do agree that this should be a warning and not an error.



> This is not eslint responsibility to try to guess that something may not work as expected. This is the reponsibility of the test suite... Why trying to guess things that are anyway checked _for sure_ by the test suite?

In my experience, most testers generally do not write test code to find race conditions—most automated tests I've seen test promises with a procedural mindset. This type of issue is generally far from "_for sure_" tested by the test suite.


> It is my understanding that `await` prevents any further execution from happening, **within that code block** until the promise is resolved. `someObj` is an argument passed into that function by reference and `someObj` is accessible everywhere it is used within that code block originally posted.

This is wrong. `await` prevents any further execution from happening, **within** _that invocation of_ **that code block** until the promise is resolved. `someObj` is an argument passed into that function by reference and `someObj` is accessible everywhere it is used within that code block _and also everywhere else that has a reference to it_.

This means that while waiting for the promise to resolve something else can modify `someObj.bar` and that by then setting it later, you are overriding that modification and it is unclear if it is the intention of the code to ignore those other potential changes.

**Comment by andiby** [comment]
You can make an other rule to handle all possible race conditions, but this rule has the name: **require-atomic-updates**.
Please read the [documentation](https://eslint.org/docs/rules/require-atomic-updates) for this rule
> - A variable or property is reassigned to a new value which is **based on its old value**.
> - A yield or await expression interrupts the assignment after the old value is read, and before the new value is set.

@stinovlas already mentioned this:
> `someObj.bar = "whatever";` assigns string `"whatever"` which is not depending on `someObj.bar` in any way.

Because of the last code changes this rule throws 4 errors for the **Examples of correct code** from the documentation:
```javascript
let result;
async function foo() {
  result = await somethingElse + result;

  let tmp = await somethingElse;
  result += tmp; // wrong error: "Possible race condition: `result` might be reassigned based on an outdated value of `result`."

  let localVariable = 0;
  localVariable += await somethingElse;
}

function* bar() {
  result += yield; // wrong error: "Possible race condition: `result` might be reassigned based on an outdated value of `result`."

  result = (yield somethingElse) + result; // wrong error: "Possible race condition: `result` might be reassigned based on an outdated value of `result`."

  result = doSomething(yield somethingElse, result); // wrong error: "Possible race condition: `result` might be reassigned based on an outdated value of `result`."
}
```
[Demo link](https://eslint.org/demo#eyJ0ZXh0IjoibGV0IHJlc3VsdDtcbmFzeW5jIGZ1bmN0aW9uIGZvbygpIHtcbiAgcmVzdWx0ID0gYXdhaXQgc29tZXRoaW5nRWxzZSArIHJlc3VsdDtcblxuICBsZXQgdG1wID0gYXdhaXQgc29tZXRoaW5nRWxzZTtcbiAgcmVzdWx0ICs9IHRtcDsgLy8gd3JvbmcgZXJyb3I6IFwiUG9zc2libGUgcmFjZSBjb25kaXRpb246IGByZXN1bHRgIG1pZ2h0IGJlIHJlYXNzaWduZWQgYmFzZWQgb24gYW4gb3V0ZGF0ZWQgdmFsdWUgb2YgYHJlc3VsdGAuXCJcblxuICBsZXQgbG9jYWxWYXJpYWJsZSA9IDA7XG4gIGxvY2FsVmFyaWFibGUgKz0gYXdhaXQgc29tZXRoaW5nRWxzZTtcbn1cblxuZnVuY3Rpb24qIGJhcigpIHtcbiAgcmVzdWx0ICs9IHlpZWxkOyAvLyB3cm9uZyBlcnJvcjogXCJQb3NzaWJsZSByYWNlIGNvbmRpdGlvbjogYHJlc3VsdGAgbWlnaHQgYmUgcmVhc3NpZ25lZCBiYXNlZCBvbiBhbiBvdXRkYXRlZCB2YWx1ZSBvZiBgcmVzdWx0YC5cIlxuXG4gIHJlc3VsdCA9ICh5aWVsZCBzb21ldGhpbmdFbHNlKSArIHJlc3VsdDsgLy8gd3JvbmcgZXJyb3I6IFwiUG9zc2libGUgcmFjZSBjb25kaXRpb246IGByZXN1bHRgIG1pZ2h0IGJlIHJlYXNzaWduZWQgYmFzZWQgb24gYW4gb3V0ZGF0ZWQgdmFsdWUgb2YgYHJlc3VsdGAuXCJcblxuICByZXN1bHQgPSBkb1NvbWV0aGluZyh5aWVsZCBzb21ldGhpbmdFbHNlLCByZXN1bHQpOyAvLyB3cm9uZyBlcnJvcjogXCJQb3NzaWJsZSByYWNlIGNvbmRpdGlvbjogYHJlc3VsdGAgbWlnaHQgYmUgcmVhc3NpZ25lZCBiYXNlZCBvbiBhbiBvdXRkYXRlZCB2YWx1ZSBvZiBgcmVzdWx0YC5cIlxufSIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJyZXF1aXJlLWF0b21pYy11cGRhdGVzIjoyfSwiZW52Ijp7ImJyb3dzZXIiOmZhbHNlLCJub2RlIjpmYWxzZSwiY29tbW9uanMiOmZhbHNlLCJzaGFyZWQtbm9kZS1icm93c2VyIjpmYWxzZSwid29ya2VyIjpmYWxzZSwiYW1kIjpmYWxzZSwibW9jaGEiOmZhbHNlLCJqYXNtaW5lIjpmYWxzZSwiamVzdCI6ZmFsc2UsInBoYW50b21qcyI6ZmFsc2UsImpxdWVyeSI6ZmFsc2UsInF1bml0IjpmYWxzZSwicHJvdG90eXBlanMiOmZhbHNlLCJzaGVsbGpzIjpmYWxzZSwibWV0ZW9yIjpmYWxzZSwibW9uZ28iOmZhbHNlLCJwcm90cmFjdG9yIjpmYWxzZSwiYXBwbGVzY3JpcHQiOmZhbHNlLCJncmVhc2Vtb25rZXkiOmZhbHNlLCJlczYiOmZhbHNlLCJ3ZWJleHRlbnNpb25zIjpmYWxzZSwiZW1iZXJ0ZXN0IjpmYWxzZSwiYXRvbXRlc3QiOmZhbHNlLCJzZXJ2aWNld29ya2VyIjpmYWxzZSwibmFzaG9ybiI6ZmFsc2V9fX0=)

So please revert the code that it works again according to the documentation.

**Comment by cburkins** [comment]
Agree, this rule seems to be a bit overly-aggressive to provide good value, disabling as well.

**Comment by glen-84** [comment]
Auto-closing an accepted issue with over 100 votes. 😕

**Comment by kaicataldo** [comment]
Reopening since this is an accepted bug. Assigning to myself so this doesn't get auto closed again. We still need to figure out how we want to move forward here.

**Comment by Lonniebiz** [comment]
To me, this issue seem very similar to one I'm starting to see with the `no-unused-vars` rule:
https://github.com/eslint/eslint/issues/13227

**Comment by denis-sokolov** [comment]
I would like to speak up in support of the rule. Mutation and async programming is a source of many errors in the codebases I had worked on and this rule has been very beneficial to bring them to light. I would also like to speak up in favor of keeping it in recommended ruleset, to help expose these difficult issues for less experienced contributors. The false positives are, of course, a major issue, as it leads to the rule being disabled or ignored.

Perhaps we can introduce more ways to explicitly mark up objects that we are allowed to mutate? This can already be done somewhat using renaming a variable in one’s scope, but it looks like a hack, since at the end of the day the object maintains identity:

```js
app.get('/', async (ctx) => {
  const myCtx = ctx;
  await process(myCtx.request.body);
  myCtx.body = 'OK';
});
```

An explicit comment could indicate the author’s intent and judgement. At this stage it is almost identical to `eslint-disable-next-line require-atomic-updates`, but with a better communication of the reasons (and it works for the entire block):

```js
app.get('/', async (ctx) => {
  // eslint-we-own: ctx
  await process(ctx.request.body);
  ctx.body = 'OK';
});
```

For common patterns like the Koa context, it would be convenient to define the variables by name:

```
```

**Comment by denis-sokolov** [comment]
I would like to speak up in support of the rule. Mutation and async programming is a source of many errors in the codebases I had worked on and this rule has been very beneficial to bring them to light. I would also like to speak up in favor of keeping it in recommended ruleset, to help expose these difficult issues for less experienced contributors. The false positives are, of course, a major issue, as it leads to the rule being disabled or ignored.

Perhaps we can introduce more ways to explicitly mark up objects that we are allowed to mutate? This can already be done somewhat using renaming a variable in one’s scope, but it looks like a hack, since at the end of the day the object maintains identity:

```js
app.get('/', async (ctx) => {
  const myCtx = ctx;
  await process(myCtx.request.body);
  myCtx.body = 'OK';
});
```

An explicit comment could indicate the author’s intent and judgement. At this stage it is almost identical to `eslint-disable-next-line require-atomic-updates`, but with a better communication of the reasons (and it works for the entire block):

```js
app.get('/', async (ctx) => {
  // eslint-we-own: ctx
  await process(ctx.request.body);
  ctx.body = 'OK';
});
```

For common patterns like the Koa context, it would be convenient to whitelist the variables we are allowed to mutate by name in the configuration:

```json
{
  "require-atomic-updates": ["error", {
    "skip-variables": ["ctx"]
  }]
}
```

**Comment by gwardwell** [comment]
@denis-sokolov I agree that the spirit of this rule And what it’s trying to solve is good. The false positives and the variables the programmer can’t control (like ctx) are major problems that must be solved. Until they are solved, I don’t believe this rule should be recommended.

**Comment by ololoepepe** [comment]
@denis-sokolov There are too many situations already, when people are treated as idiots ("for their own safety", of course). Do not create another one, please.

**Comment by mdjermanovic** [comment]
I think we should add an option to ignore assignments to properties. That is, to report only assignments to variables if this option is enabled.

**Comment by nzakas** [comment]
> I think we should add an option to ignore assignments to properties. That is, to report only assignments to variables if this option is enabled.

Agree. I’m not sure the current default actually makes sense but an option is an easy way to address it.
